nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py

```diff
@@ -0,0 +1,290 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from copy import deepcopy
+import logging
+from typing import Dict, List, Tuple, Callable
+
+import torch
+from torch import autograd, Tensor
+from torch.nn import Module, Parameter
+from torch.optim import Optimizer, Adam
+
+from nni.algorithms.compression.v2.pytorch.base.compressor import Compressor, _setattr, LayerInfo
+from nni.algorithms.compression.v2.pytorch.pruning.basic_pruner import BasicPruner, NORMAL_SCHEMA, EXCLUDE_SCHEMA, INTERNAL_SCHEMA
+from nni.algorithms.compression.v2.pytorch.utils import CompressorSchema
+
+from .tools.base import TrainerBasedDataCollector
+
+from .tools import (
+    StraightMetricsCalculator,
+    NormalSparsityAllocator
+)
+
+_logger = logging.getLogger(__name__)
+
+
+class PrunerScoredModuleWrapper(Module):
+    """
+    Wrap an module to enable data parallel, forward method customization and buffer registeration.
+    Different from `PrunerModuleWrapper`, `PrunerScoredModuleWrapper` will record the gradient.
+
+    Parameters
+    ----------
+    module
+        The module user wants to compress.
+    config
+        The configurations that users specify for compression.
+    module_name
+        The name of the module to compress, wrapper module shares same name.
+    pruner
+        The pruner used to calculate mask.
+    """
+    def __init__(self, module: Module, module_name: str, config: Dict, pruner: Compressor):
+        super().__init__()
+        # origin layer information
+        self.module = module
+        self.name = module_name
+        # config and pruner
+        self.config = config
+        self.pruner = pruner
+
+        self.weight = Parameter(torch.empty(self.module.weight.size()))
+        self.weight.data = self.module.weight.data
+
+        self.weight_score = Parameter(torch.empty(self.weight.size()))
+        torch.nn.init.constant_(self.weight_score, val=0.0)
+
+        # register buffer for mask
+        self.register_buffer("weight_mask", torch.ones(self.module.weight.shape))
+        if hasattr(self.module, 'bias') and self.module.bias is not None:
+            self.register_buffer("bias_mask", torch.ones(self.module.bias.shape))
+            self.bias = Parameter(torch.empty(self.module.bias.size()))
+            self.bias.data = self.module.bias.data
+        else:
+            self.register_buffer("bias_mask", None)
```

Comment:
if there is no bias, do we need to create bias_mask?

Reply:
In fact, we don't need it, but we want to be compatible with the old version.